econml/dml_iv.py

```diff
@@ -0,0 +1,760 @@
+        Possible inputs for cv are:
+
+        - None, to use the default 3-fold cross-validation,
+        - integer, to specify the number of folds.
+        - :term:`CV splitter`
+        - An iterable yielding (train, test) splits as arrays of indices.
+
+        For integer/None inputs, if the treatment is discrete
+        :class:`~sklearn.model_selection.StratifiedKFold` is used, else,
+        :class:`~sklearn.model_selection.KFold` is used
+        (with a random shuffle in either case).
+
+        Unless an iterable is used, we call `split(concat[W, X], T)` to generate the splits. If all
+        W, X are None, then we call `split(ones((T.shape[0], 1)), T)`.
+
+    random_state: int, :class:`~numpy.random.mtrand.RandomState` instance or None, optional (default=None)
+        If int, random_state is the seed used by the random number generator;
+        If :class:`~numpy.random.mtrand.RandomState` instance, random_state is the random number generator;
+        If None, the random number generator is the :class:`~numpy.random.mtrand.RandomState` instance used
+        by :mod:`np.random<numpy.random>`.
+    """
+
+    def __init__(self,
+                 nuisance_models,
+                 model_final,
+                 cov_clip=0.1, opt_reweighted=False,
+                 discrete_instrument=False, discrete_treatment=False,
+                 n_splits=2, random_state=None):
+        class ModelFinal:
+            """
+            Final model at fit time, fits a residual on residual regression with a heterogeneous coefficient
+            that depends on X, i.e.
+
+                .. math ::
+                    Y - E[Y | X] = \\theta(X) \\cdot (E[T | X, Z] - E[T | X]) + \\epsilon
+
+            and at predict time returns :math:`\\theta(X)`. The score method returns the MSE of this final
+            residual on residual regression.
+            """
+
+            def __init__(self, model_final):
+                self._model_final = clone(model_final, safe=False)
+
+            @staticmethod
+            def _effect_estimate(nuisances):
+                prel_theta, res_t, res_y, res_z, cov = nuisances
+
+                # Estimate final model of theta(X) by minimizing the square loss:
+                # (prel_theta(X) + (Y_res - prel_theta(X) * T_res) * Z_res / cov[T,Z | X] - theta(X))^2
+                # We clip the covariance so that it is bounded away from zero, so as to reduce variance
+                # at the expense of some small bias. For points with very small covariance we revert
+                # to the model-based preliminary estimate and do not add the correction term.
+                cov_sign = np.sign(cov)
+                cov_sign[cov_sign == 0] = 1
+                clipped_cov = cov_sign * np.clip(np.abs(cov),
+                                                 self.cov_clip, np.inf)
+                return prel_theta + (res_y - prel_theta * res_t) * res_z / clipped_cov
```

Comment:
This eq also assumes single Y and T. This should be documented and an exception should be thrown when the user passes in objects that cannot be flattened.

Reply:
Done